grpc/creds/creds.go

```diff
@@ -51,9 +63,73 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
 	}
 }
 
+func (tc *transportCredentials) peerIsWhitelisted(peerState tls.ConnectionState) error {
+	// If there's no whitelist, all clients are OK
+	if tc.whitelist == nil {
+		return nil
+	}
+
+	// Otherwise its time to start inspecting the peer's `VerifiedChains`
+	chains := peerState.VerifiedChains
+	if len(chains) < 1 {
```

Comment:
Also, looking at this again, every entry in `VerifiedChains` should have the same zeroth entry, so the loop isn't really necessary. What happens if the client certificate doesn't verify? Do we wind up here with `PeerCertificates` set, but `VerifiedChains` empty? Or do we just never get here if no trusted path can be built? Unfortunately https://golang.org/pkg/crypto/tls/#ConnectionState doesn't really explain the semantics, so this may require checking the source (and possibly filing a doc bug on Go).